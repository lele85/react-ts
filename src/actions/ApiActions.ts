export enum ApiActionsKeys {
    REQUEST = "@@api/FETCH_REQUEST",
    SUCCESS = "@@api/FETCH_SUCCESS",
    ERROR = "@@api/FETCH_ERROR",
};

type ApiRequestParams<T> = {
    [P in keyof T] : T[P]
};

export interface ApiFetchRequestAction<Tparams> {
    type: ApiActionsKeys.REQUEST,
    params: ApiRequestParams<Tparams>
};

export interface ApiFetchSuccessAction<Tparams, Tmodel> {
    type: ApiActionsKeys.SUCCESS,
    params: ApiRequestParams<Tparams>,
    model: Tmodel
};

export interface ApiFetchErrorAction<Tparams> {
    type: ApiActionsKeys.ERROR,
    params: ApiRequestParams<Tparams>
};