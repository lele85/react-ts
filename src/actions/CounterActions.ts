import actionCreatorFactory from 'typescript-fsa';

const actionCreator = actionCreatorFactory("@@counter");
export const counterInc = actionCreator<{by: number}>('INC');
export const counterDec = actionCreator<{by: number}>('DEC');