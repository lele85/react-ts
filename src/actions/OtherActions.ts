export enum OtherActionsKeys {
	OTHER_ACTION = "@@redux/OTHER_ACTION",
};

export interface OtherAction {
    type: OtherActionsKeys.OTHER_ACTION;
};