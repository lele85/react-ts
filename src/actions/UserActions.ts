import actionCreatorFactory from 'typescript-fsa';

const actionCreator = actionCreatorFactory("@@user");
export const userSetFirstname = actionCreator<{firstname: string}>('SET_FIRSTNAME');
export const userSetLastname = actionCreator<{lastname: string}>('SET_LASTNAME');