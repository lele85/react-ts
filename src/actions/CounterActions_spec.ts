import { CounterState } from '../state/CounterState';
import { createMockStore } from '../lib/TestUtils';
import { counterInc, counterDec } from './CounterActions';

describe("Counter Actions", () => {

    it("should be dispatched", async () => { 
        const store = createMockStore<CounterState>({value:0});
        store.dispatch(counterInc({by:2}));
        store.dispatch(counterDec({by:3}));
        store.dispatch(counterInc({by:10}))
        expect(store.getActions()).toMatchSnapshot();
    });

});