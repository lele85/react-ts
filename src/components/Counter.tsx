import React from 'react';
import { StatelessComponent } from 'react';
import { connect } from 'react-redux';
import { Dispatch } from 'redux';
import { bindActionCreators } from 'redux';
import { AnyAction } from 'typescript-fsa';
import { $call } from 'utility-types';

import { counterDec, counterInc } from '../actions/CounterActions';
import { AppState } from '../state/AppState';
import styles from './Counter.module.css';

const mapStateToProps = (state:AppState) => {
    return {
        counter: state.counter.value
    };
}

const mapActionsToProps = (dispatch : Dispatch<AnyAction>) => {
    return bindActionCreators({
        increment: counterInc,
        decrement: counterDec
    }, dispatch);
}

const returnOfActionProps = $call(mapActionsToProps);
const returnOfStateProps = $call(mapStateToProps);
type ActionProps = typeof returnOfActionProps;
type StateProps = typeof returnOfStateProps;
type Props = StateProps & ActionProps;

export const Counter: StatelessComponent<Props> = ({counter, increment, decrement}) => {
    return (
        <h1 className={styles.counter}>
            <div>{counter}</div>
            <button onClick={() => { increment({by:1}); }}>+1</button>
            <button onClick={() => { decrement({by:1}); }}>-1</button>
        </h1>
    );
}

export default connect(mapStateToProps, mapActionsToProps)(Counter);