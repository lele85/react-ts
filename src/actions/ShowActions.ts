import { ApiFetchRequestAction, ApiFetchSuccessAction, ApiFetchErrorAction, ApiActionsKeys, fetchApi } from "./ApiActions";
import { ShowsParams, Show } from "../model/Show";
import { OtherAction } from "./OtherActions";
import { Dispatch } from "redux";
import { get } from "../lib/Http";


export type FetchShowsRequestActionTypes =
   | ApiFetchRequestAction<ShowsParams>
   | ApiFetchSuccessAction<ShowsParams, Array<Show>>
   | ApiFetchErrorAction<ShowsParams>
   | OtherAction;

export const fetchShows = () => {
    return fetchApi<ShowsParams, Show[], FetchShowsRequestActionTypes>("/api/shows", {});
};