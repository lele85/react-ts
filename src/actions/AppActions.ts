import { Dispatch } from "react-redux";
import { CounterActionTypes } from "./CounterActions";
import { UserActionTypes } from "./UserActions";
import { OtherAction } from "./OtherActions";

export type AppActionTypes =
    | CounterActionTypes
    | UserActionTypes
    | OtherAction;