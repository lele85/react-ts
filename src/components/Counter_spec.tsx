import { shallow } from 'enzyme';
import React from 'react';
import { bindActionCreators } from 'redux';

import { counterDec, counterInc } from '../actions/CounterActions';
import { Counter } from './Counter';

it("renders correctly", () => {
    const {increment,decrement} = bindActionCreators({
        increment: counterInc,
        decrement: counterDec
    }, () => {});

    const result = shallow(
        <Counter
            counter={0}
            increment={increment}
            decrement={decrement}
        />
    );
    expect(result).toMatchSnapshot();
});