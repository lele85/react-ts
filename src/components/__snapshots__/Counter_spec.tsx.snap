// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Unknown
    counter={0}
    decrement={[Function]}
    increment={[Function]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div>
          0
        </div>,
        <button
          onClick={[Function]}
        >
          +1
        </button>,
        <button
          onClick={[Function]}
        >
          -1
        </button>,
      ],
      "className": "counter",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": 0,
        },
        "ref": null,
        "rendered": 0,
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "+1",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "+1",
        "type": "button",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "-1",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "-1",
        "type": "button",
      },
    ],
    "type": "h1",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div>
            0
          </div>,
          <button
            onClick={[Function]}
          >
            +1
          </button>,
          <button
            onClick={[Function]}
          >
            -1
          </button>,
        ],
        "className": "counter",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": 0,
          },
          "ref": null,
          "rendered": 0,
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "+1",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "+1",
          "type": "button",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "-1",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "-1",
          "type": "button",
        },
      ],
      "type": "h1",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
